### ISSUE TITLE
Find ways to minimize the effort required to configure authz in an HTTP API by 30%

### ISSUE THREAD

**Issue by bradygaster** [dotnet/aspnetcore#42192]
[This item was copied from the planning repo]

Through Cu, work with partners in AAD and internally within ASP.NET to collaborate on user research and experience measurement to find ways we can ease the burden of configuring popular auth providers like Microsoft Identity Platform or and 2 popular 3rd party providers. 

Specifically concentrate on easing the friction and effort (keystrokes, clicks, and internet searches required) for one or all of the following scenarios:

- [ ] Given a Swashbuckle-enabled Web or Minimal API project, add support for authenticating the user on the Swagger UI page. 
- [ ] Adding authz to an existing ASP.NET Core API so that it can be called only from an authenticated client.
- [ ] Assessing if we need to add support for the OAuth2 endpoints for Twitter (our docs specifically call out that a 3rd party NuGet package can be used for this, but that package fails to work with ASP.NET Core identity, so customers who want to support Twitter auth are in a broken state; they lack support for the OAuth2 features with our provider but can't authenticate customers with the OSS provider. 
- [ ] Asses how many of the authz-related docs specific to social or 3rd-party auth providers in our ASP.NET Core Identity docs section can be completed by PMs on the team according to the docs in [this TOC node](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/social/?view=aspnetcore-6.0&tabs=visual-studio) and establish a content plan, if needed, for resolving hurdles in the experience. 

As we improve the scenario with product work, we should accompany that product work with documentation for our testing team so the experience of configuring auth for these key scenarios can be tested repeatedly and so we know when service or API changes result in our needing to update the docs or the experience.

**Comment by kevinchalet** [comment]
I just tested locally and I haven't encountered any issue with ASP.NET Core Identity.

@martincostello do you mind giving it a try to confirm it works fine for you too?

**Comment by bradygaster** [comment]
Oh i apologize - just saw this. :) In fact I tried with both the one we ship and the one from your side, in an experiment with Blazor  WebAssembly auth. I hadn't filed an issue against your repo because I hadn't validated it wasn't my own code doing something awry. I'd love to sync up with you and show you what I ran into when I tried your libs.

**Comment by martincostello** [comment]
Blazor is an untested scenario on our side, mainly through lack of experience. We often get questions about it though for our OAuth providers in general (not specifically Twitter).

If it’s a scenario that _should_ work with Blazor and it’s fairly trivial to fix whatever’s wrong, then I think we’d be happy to work with you and the team to light it up.

**Comment by bradygaster** [comment]
Absolutely appreciate that @martincostello and @kevinchalet - I've used a few of your other providers for projects over the years so I was eager when the docs pointed me in your direction. Once you see my code/attempt-at-code, if we find something I can tweak, I'd be eager to update the doc with something more tangible than a link showing folks who want to use the oauth2 approaches you provide that we don't, so I'd be stoked to work on a deeper article highlighting some of what's possible with it once we figure out what else I've done wrong in the sample.

**Comment by martincostello** [comment]
I can't speak for Kévin's availability, but I'd be happy to do a call with you 1600-1800 BST (0800-1000 PDT?) any weekday to try and work through what you have and do some troubleshooting.

You can get some [contact details](https://github.com/martincostello#contact) off my GH profile.

**Comment by kevinchalet** [comment]
@bradygaster sent me a repro: it's a classical Blazor WASM+Identity Server app with a custom `ExternalLogin.cshtml` that uses the popular Tweetinvi library to retrieve the profile image and the email address (in this case, the Twitter authentication part is purely handled server-side and the Blazor WASM client gets local tokens from IdSrv after a transparent authorization dance).

There are 2 distinct issues:
  - The Tweetinvi API used to achieve that - `Users.GetAuthenticatedUserAsync()` - targets the v1 API endpoints. Yet, these endpoints don't work with the Twitter OAuth 2.0 integration, that requires using the v2 endpoints. Once you use the `UsersV2.GetUserByNameAsync("me")` API to get the profile image (AFAIK, you can't retrieve the email address using the v2 user endpoints), things work fine.

  - From time to time, too much data is stored in the `AuthenticationProperties` and the handler ends up generating a `state` bag that is too large for Twitter. Empirically, I was able to determine that states longer than 500 chars cause a generic error. I'm fairly sure this potential issue was discussed during the review. Fixing that won't be exactly trivial as it will require storing the state elsewhere (e.g in a distributed cache or in a database, as the OpenIddict client does)

**Comment by martincostello** [comment]
Ah cool, glad one of them was a fairly easy fix.

The second issue sounds like the problem I had with the first implementation that _seemed_ to fix itself: https://github.com/dotnet/aspnetcore/issues/39664#issuecomment-1018598236

**Comment by bradygaster** [comment]
Oh that's very interesting - adding @ReubenBond as you said "distributed caching" - an idea he and the Orleans team and I are discussing.

**Comment by bradygaster** [comment]
@kevinchalet - to be clear, I only used the Tweetinvi client because I couldn't figure out how to use the claims to get all the properties directly from Twitter. If I could make that work, I'd probably have no use for that client package.

**Comment by martincostello** [comment]
I'm not at a computer right now to try it myself by running the code, but maybe there's addition fields you can request in the user payload, and then map those to claims?

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/docs/twitter.md

**Comment by kevinchalet** [comment]
_Crap, @martincostello is too fast for me_ 🤣

**Comment by martincostello** [comment]
I haven't looked at getting the email yet, but adding these [two lines of configuration](https://github.com/martincostello/AspNet.Security.OAuth.Providers/commit/bc1069cd9120fb46785710d818569fdc86aadcee) for the provider using [our repo's sample app](https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/tree/dev/samples/Mvc.Client) gets me a claim containing the profile image's URL.

```csharp
.AddTwitter(options =>
{
    options.ClientId = Configuration["Twitter:ClientId"];
    options.ClientSecret = Configuration["Twitter:ClientSecret"];
    options.UserFields.Add("profile_image_url");
    options.ClaimActions.MapJsonSubKey("urn:twitter:avatar", "data", "profile_image_url");
})
```

![image](https://user-images.githubusercontent.com/1439341/174029172-1e0eb2bb-2033-4a79-8aa0-15f0ca52bc64.png)

**Comment by martincostello** [comment]
It appears that it isn't possible to get the user's email address using the v2 Twitter API: [twittercommunity.com](https://twittercommunity.com/t/how-to-get-user-details-by-email-address-is-there-any-api-exists/143073)

**Comment by martincostello** [comment]
I haven't run into the _"state too long"_ problem at all locally with our sample - probably because we don't do anything additional with the `AuthenticationProperties`.

I'll have a think about how we could provide a way to work around that that isn't tied to much to a specific solution and/or is extensible.

**Comment by bradygaster** [comment]
> It appears that it isn't possible to get the user's email address using the v2 Twitter API: [twittercommunity.com](https://twittercommunity.com/t/how-to-get-user-details-by-email-address-is-there-any-api-exists/143073)

Given ASP.NET identity requires email, this seems bad. Any workarounds you'd advise?

**Comment by martincostello** [comment]
I’ll give it a try tomorrow, but I _assumed_ not as it’s part of the 1.1 API.

**Comment by martincostello** [comment]
Nope, that doesn't work. Using the access token as-is gets you an HTTP 403 error.

I looked at the code for `ExecuteRequestAsync()`, but that needs a token secret to generate the relevant signature to call the API which you don't have in the OAuth 2 flow.

Assuming I've understood it all correctly, you can only get the email by using the ASP.NET Core provider that uses the 1.1 API. Otherwise you have to mix and match the two API versions to use the different capabilities.

The snippet I added to test the above was this:

```csharp
options.Events.OnCreatingTicket = async context =>
{
    context.Backchannel.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue(
        "Bearer",
        context.AccessToken);

    using var response = await context.Backchannel.GetAsync(
        "https://api.twitter.com/1.1/account/verify_credentials.json?include_email=true",
        context.HttpContext.RequestAborted);

    response.EnsureSuccessStatusCode();

    using var content = await response.Content.ReadAsStreamAsync();
    using var document = await JsonDocument.ParseAsync(content);

    if (document.RootElement is { } root)
    {
        // Do stuff
    }
};
```

According to the [Twitter documentation](https://developer.twitter.com/en/docs/twitter-api/v1/accounts-and-users/manage-account-settings/api-reference/get-account-verify_credentials) for the endpoint there's no guarantee an email would be returned so even with the ASP.NET Core Twitter provider an application using ASP.NET Core Identity would need to handle the case of their being no email address when the user logs in:

> If the user does not have an email address on their account, or if the email address is not verified, null will be returned.

**Comment by bradygaster** [comment]
> Assuming I've understood it all correctly, you can only get the email by using the ASP.NET Core provider that uses the 1.1 API. Otherwise you have to mix and match the two API versions to use the different capabilities.

This was my presumption at first glance. Eesh. Kinda sub-optimal.

**Comment by kevinchalet** [comment]
@bradygaster not sure if it works for you, but you always can ask the users for their email address using a good old form 😃 (it's what the default Identity UI does).

**Comment by adityamandaleeka** [comment]
@bradygaster Can you put this issue into a milestone please?
